src/System.Text.Json/src/System/Text/Json/Reader/Utf8JsonReader.MultiSegment.cs

```diff
@@ -2449,207 +2332,242 @@ private bool ConsumeSingleLineCommentMultiSegment(ReadOnlySpan<byte> localBuffer
                 {
-                    idx = localBuffer.Length;
-                    toConsume = idx;
-                    // Assume everything on this line is a comment and there is no more data.
-                    _bytePositionInLine += 2 + localBuffer.Length;
-                    goto Done;
+                    if (expectLF)
+                    {
+                        break;
+                    }
+
+                    return true;
                 }
 
                 if (!GetNextSpan())
                 {
-                    _totalConsumed = prevTotalConsumed;
-                    return false;
+                    if (IsLastSpan)
+                    {
+                        if (expectLF)
+                        {
+                            break;
+                        }
+
+                        return true;
+                    }
+                    else
+                    {
+                        return false;
+                    }
                 }
-                HasValueSequence = true;
-                _totalConsumed += localBuffer.Length + leftOver;
-                previousConsumed = 0;
-                leftOver = 0;
+
                 localBuffer = _buffer;
-                idx = -1;
-            } while (true);
+            }
 
-            idx++;
             _bytePositionInLine = 0;
             _lineNumber++;
 
-        Done:
-            if (HasValueSequence)
+            return true;
+        }
+
+        private int FindLineSeparatorMultiSegment(ReadOnlySpan<byte> localBuffer, ref int dangerousLineSeparatorBytesConsumed)
+        {
+            Debug.Assert(dangerousLineSeparatorBytesConsumed >= 0 && dangerousLineSeparatorBytesConsumed <= 2);
+
+            if (dangerousLineSeparatorBytesConsumed != 0)
             {
-                ReadOnlySequence<byte> commentSequence = _sequence.Slice(start, end);
-                if (commentSequence.IsSingleSegment)
+                ThrowOnDangerousLineSeparatorMultiSegment(localBuffer, ref dangerousLineSeparatorBytesConsumed);
+
+                if (dangerousLineSeparatorBytesConsumed != 0)
                 {
-                    ValueSpan = commentSequence.First.Span;
-                    HasValueSequence = false;
+                    // this can only happen if localBuffer size is 1 and we have previously only consumed 1 byte
+                    // or localBuffer is 0
+                    Debug.Assert(dangerousLineSeparatorBytesConsumed >= 1 && dangerousLineSeparatorBytesConsumed <= 2 && localBuffer.Length <= 1);
+                    return -1;
                 }
-                else
+            }
+
+            int totalIdx = 0;
+            while (true)
+            {
+                int idx = localBuffer.IndexOfAny(JsonConstants.LineFeed, JsonConstants.CarriageReturn, JsonConstants.StartingByteOfNonStandardSeparator);
+                dangerousLineSeparatorBytesConsumed = 0;
+
+                if (idx == -1)
+                    return -1;
```

Comment:
nit: add braces here/elsewhere (whenever you have an if within a nested scope, even if it happens to be a single line).